pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -146,65 +148,175 @@ func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncCo
+	// work, err = applyWork(ctx, c.workClient, c.workLister, c.cache, c.eventRecorder, work)
+	// if err != nil {
+	// 	meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+	// 		Type:    "ManifestApplied",
+	// 		Status:  metav1.ConditionFalse,
+	// 		Reason:  "ManifestWorkApplyFailed",
+	// 		Message: fmt.Sprintf("failed to apply manifestwork: %v", err),
+	// 	})
+	// 	if _, updateErr := c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+	// 		ctx, managedClusterAddonCopy, metav1.UpdateOptions{}); err != nil {
+	// 		return fmt.Errorf("failed to update managedclusteraddon status: %v; the err should be %v", updateErr, err)
+	// 	}
+	// 	return err
+	// }
+
+	// // Update addon status based on work's status
+	// if meta.IsStatusConditionTrue(work.Status.Conditions, workapiv1.WorkApplied) {
+	// 	meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+	// 		Type:    "ManifestApplied",
+	// 		Status:  metav1.ConditionTrue,
+	// 		Reason:  "AddonManifestApplied",
+	// 		Message: "manifest of addon applied successfully",
+	// 	})
+	// } else {
+	// 	meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+	// 		Type:    "ManifestApplied",
+	// 		Status:  metav1.ConditionFalse,
+	// 		Reason:  "AddonManifestAppliedFailed",
+	// 		Message: fmt.Sprintf("work %s apply failed", work.Name),
+	// 	})
+	// }
+
+	if equality.Semantic.DeepEqual(managedClusterAddonCopy.Status, managedClusterAddon.Status) {
+		return nil
+	}
+
+	_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+		ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+	return err
+}
+
+func (c *addonDeployController) applyManifestWork(ctx context.Context, agentAddon agent.AgentAddon,
+	managedClusterAddon *addonapiv1alpha1.ManagedClusterAddOn,
+	managedCluster *clusterv1.ManagedCluster,
+	manifestworkNamespace string,
+	conditionTypePrefix string,
+	getManifests func(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error)) error {
+
+	owner := metav1.NewControllerRef(managedClusterAddon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddOn"))
+
+	objects, err := getManifests(managedCluster, managedClusterAddon.DeepCopy())
 	if err != nil {
 		return err
 	}
 	if len(objects) == 0 {
 		return nil
 	}
 
-	work, _, err := buildManifestWorkFromObject(clusterName, addonName, objects)
+	work, _, err := buildManifestWorkFromObject(manifestworkNamespace, managedClusterAddon.Name, objects)
 	if err != nil {
 		return err
 	}
-	work.OwnerReferences = []metav1.OwnerReference{*owner}
+
+	// FIXME: the ManagedClusterAddon can NOT be the owner as management manifests(different namespace), delete them
+	// manually when the ManagedClusterAddon is deleted.
+	if len(conditionTypePrefix) == 0 {
+		work.OwnerReferences = []metav1.OwnerReference{*owner}
```

Comment:
this needs a finalizer on addon resource. using conditionTypePrefix to check is also tricky.

Reply:
yes, this is just for a test and should be fixed.